### ISSUE TITLE
Some numeric tests from TSPL4 don't match IronScheme results.

### ISSUE THREAD

**Issue by bcalco** [IronScheme/IronScheme#110]
Hi @leppie,

Some of these may just be niggling nuances, but my tester was told to use TSPL4 examples as the baseline and so I'm just raising these (in some cases minor) discrepancies in the interest of correctness:

From examples [here](https://www.scheme.com/tspl4/objects.html#./objects:s110):

```
> (equal? (expt +i 2) -1)
#f
> (= (expt +i 2) -1)
#f
> (expt +i 2)
-1.0+1.2246063538223773e-16i
```
From examples [here](https://www.scheme.com/tspl4/objects.html#./objects:s125):
```
> (equal? (sqrt -3.0-4.0i) '1.0-2.0i) 
#f
> (= (sqrt -3.0-4.0i) '1.0-2.0i) ;;; should return #t according to TSPL4.
#f
> (sqrt -3.0-4.0i)
1.0000000000000002-2.0i
>
```

From examples [here](https://www.scheme.com/tspl4/objects.html#./objects:s148):

```
> (equal? (number->string 1e-23) "1e-23")
#f
> (eqv? (number->string 1e-23) "1e-23")
#f
> (number->string 1e-23)
"1.e-23"
```
Obviously to human eyes these are equivalent, but `(eqv? ...)` doesn't agree.

From examples [here](https://www.scheme.com/tspl4/objects.html#./objects:s148):
```
> (equal? (number->string 220/9 16) "DC/9")
#f
> (eqv? (number->string 220/9 16) "DC/9")
#f
> (number->string 220/9 16)
"dc/9"
```
I note that the "DC" may not have to be case sensitive to be correct. In which case probably `(eqv? ...)` should return #t. To human eyes these are plainly equivalent.

From examples [here](https://www.scheme.com/tspl4/objects.html#./objects:s288)
```
> (define ht (make-eq-hashtable))
> (define p1 (cons 'a 'b))
> (define p2 (cons 'a 'b))
> (hashtable-set! ht p1 "one")
> (hashtable-set! ht p2 "two")
> (hashtable-set! ht 'q "three")
> (hashtable-keys ht)
#((a . b) (a . b) q)
> (equal? (hashtable-keys ht) '#(q (a . b) (a . b)))
#f
> (eqv? (hashtable-keys ht) '#(q (a . b) (a . b)))
#f
> (eq? (hashtable-keys ht) '#(q (a . b) (a . b)))
#f
```
The test is failing only because of strict order is being compared. Human eyeballs see the contents of the vector returned are the same, but the order is different. Because order shouldn't matter, IMHO `(eqv? ...)` should return true and be the appropriate equality test here. No?

I will post more of these kinds of things here as we find them.

**Comment by leppie** [comment]
Clearly Chez use a better library for complex math (GMP iirc).

You are also breaking floating point rule number 1 :D 

> -1.0+1.2246063538223773e-16i

Close enough

> 1.0000000000000002-2.0i

Close enough

Next up you are comparing string output to string output. So it does not matter what the actual values are.

> "1.e-23"

```
> 1e-23
1.e-23
> 1.e-23
1.e-23
```

Valid R6RS.  I could maybe improve it for these specific cases.

> "dc/9"

```
> #xdc/9
220/9
> #xDC/9
220/9
```

Valid R6RS. You can use either case.

> The test is failing only because of strict order is being compared.

Hashtables have no implicit order defined, the order will depend on the internal hashing algorithm used.

> IMHO (eqv? ...) should return true and be the appropriate equality test here. No? 

That wont work as vector returned in IronScheme's case is a fresh instance. R6RS does not require it be the same instance. Chez clearly optimizes this to reuse the same vector, hence why `eq?` also works. 

```
> (define v1 '#(1 2 3))
> (define v2 '#(1 2 3))
> (eq? v1 v2)
#f
> (eqv? v1 v2)
#f
> (equal? v1 v2)
#t
```

**Comment by bcalco** [comment]
Yeah, I'm not sure how to treat most of these. If you decide to update any of your formatting or other logic to more closely align with Chez's results, let me know.

**Comment by leppie** [comment]
FYI, I get this on Chez

```
> (expt +1.0i 2)
-1.0+1.2246467991473532e-16i
> (expt +i 2.0)
-1.0+1.2246467991473532e-16i
```

So it seems it has some concept of 'exact' complex numbers. And higher precision 'inexact' ones.
